File: python/telekinesis/cryptography.py
import base64
import ujson
import os
import time

from cryptography.hazmat.primitives.asymmetric import ec, utils
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.exceptions import InvalidSignature


class PrivateKey:
    def __init__(self, key_file=None, password=None):
        self._public_serial_cache = None
        if key_file and os.path.exists(key_file):
            with open(key_file, "rb") as kf:
                data = kf.read()
            self.key = PrivateKey.from_private_serial(data, password).key
            return

        self.key = ec.generate_private_key(curve=ec.SECP256R1, backend=default_backend())

        if key_file:
            self.save_key_file(key_file, password)

    def sign(self, m):
        signature = self.key.sign(m, ec.ECDSA(hashes.SHA256()))
        r, s = utils.decode_dss_signature(signature)
        return b"".join([x.to_bytes(32, "big") for x in (r, s)])

    def public_serial(self):
        if not self._public_serial_cache:
            Q = self.key.public_key().public_numbers()
            self._public_serial_cache = base64.b64encode(b"".join([x.to_bytes(32, "big") for x in (Q.x, Q.y)])).decode()
        return self._public_serial_cache

    def _private_serial(self, password=None):
        enc = (
            serialization.NoEncryption()
            if password is None
            else serialization.BestAvailableEncryption(password.encode())
        )
        return self.key.private_bytes(
            encoding=serialization.Encoding.PEM, format=serialization.PrivateFormat.PKCS8, encryption_algorithm=enc,
        )
    def save_key_file(self, key_file, password=None):
        with open(key_file, "wb") as kf:
            kf.write(self._private_serial(password))

    @staticmethod
    def from_private_serial(data, password=None):
        out = PrivateKey()
        out.key = serialization.load_pem_private_key(
            data, None if password is None else password.encode(), default_backend()
        )
        return out

    def __getstate__(self):
        return {'private_serial': self._private_serial()}

    def __setstate__(self, state):
        self.key = PrivateKey.from_private_serial(state['private_serial']).key


class PublicKey:
    def __init__(self, public_serial):
        self.key = ec.EllipticCurvePublicKey.from_encoded_point(ec.SECP256R1(), b"\x04" + base64.b64decode(public_serial))

    def verify(self, raw_signature, message):
        r = int.from_bytes(raw_signature[:32], "big")
        s = int.from_bytes(raw_signature[32:], "big")

        signature = utils.encode_dss_signature(r, s)

        self.key.verify(signature, message, ec.ECDSA(hashes.SHA256()))


class SharedKey:
    def __init__(self, private_key, public_key):
        self.key = private_key.key.exchange(ec.ECDH(), public_key.key)

    def encrypt(self, message, nonce):
        encryptor = Cipher(algorithms.AES(self.key), modes.CTR(nonce), default_backend()).encryptor()

        return encryptor.update(message) + encryptor.finalize()

    def decrypt(self, ciphertext, nonce):
        decryptor = Cipher(algorithms.AES(self.key), modes.CTR(nonce), default_backend()).decryptor()

        return decryptor.update(ciphertext) + decryptor.finalize()


class Token:
    def __init__(
        self,
        issuer,
        brokers,
        receiver,
        asset,
        token_type,
        max_depth=None,
        valid_from=None,
        valid_until=None,
        metadata=None,
    ):
        self.issuer = issuer
        self.brokers = brokers
        self.receiver = receiver
        self.asset = asset
        self.token_type = token_type
        self.max_depth = max_depth
        self.valid_from = valid_from or time.time()
        self.valid_until = valid_until
        self.metadata = metadata or {}
        self.signature = None

    def verify(self, signature):
        try:
            PublicKey(self.issuer).verify(base64.b64decode(signature.encode()), self._to_string().encode())
            self.signature = signature
            return True
        except InvalidSignature:
            return False

    def _to_dict(self):
        return {
            x: self.__getattribute__(x)
            for x in [
                "issuer",
                "brokers",
                "receiver",
                "asset",
                "token_type",
                "max_depth",
                "valid_from",
                "valid_until",
                "metadata",
            ]
        }

    def _to_string(self):
        return ujson.dumps(self._to_dict(), escape_forward_slashes=False)

    def encode(self):
        if not self.signature:
            raise RuntimeError("You need to sign the token before encoding it")
        return self.signature + "." + self._to_string()

    def sign(self, signing_key):
        self.signature = base64.b64encode(signing_key.sign(self._to_string().encode())).decode()
        return self.signature

    def __repr__(self):
        return f"{self.issuer[:4]} - {self.signature and self.signature[:4]}: {self.receiver[:4]} ({self.asset[:4]})"

    @staticmethod
    def decode(string, verify=True):
        token = Token(**ujson.loads(string[string.find(".") + 1 :]))
        if not verify:
            token.signature = string.split(".")[0]
            return token
        if token.verify(string.split(".")[0]):
            return token
        raise InvalidSignature

File: js/src/cryptography.ts
import { b64decode, b64encode } from "./utils";

const webcrypto = typeof crypto == 'undefined' ? global.crypto : crypto;

export class PrivateKey {
  _algorithm: "ECDSA" | "ECDH";
  _usage: ["sign"] | ["deriveKey"];
  repr?: string;
  key?: CryptoKeyPair;
  exportedKey?: string;
  name?: string;

  constructor(use: "sign" | "derive", importedKey?: string, name?: string) {
    this.exportedKey = importedKey;
    this.name = importedKey ? importedKey.split('-----END PRIVATE KEY-----').slice(1).join('-----END PRIVATE KEY-----').replace('\n', ''): name;
    if (use === 'sign') {
      this._algorithm = 'ECDSA';
      this._usage = ['sign'];
      return;
    }
    this._algorithm = 'ECDH';
    this._usage = ['deriveKey'];
    // this.publicSerial(); // <- don't do this!
    return
  }
  async generate() {
    if (this.exportedKey !== undefined) {
      let privateKey, jwk;
      privateKey = await webcrypto.subtle.importKey(
        "pkcs8",
        b64decode(this.exportedKey.split('-----END PRIVATE KEY-----')[0].split('\n').slice(1).join('')),
        {
          name: this._algorithm,
          namedCurve: "P-256"
        },
        true,
        this._usage
      );
      jwk = await webcrypto.subtle.exportKey("jwk", privateKey);
      this.key = {
        privateKey: privateKey,
        publicKey: await webcrypto.subtle.importKey(
          'jwk',
          {alg: jwk.alg, crv: jwk.crv, ext: true, key_ops: [], kty: jwk.kty, x: jwk.x, y: jwk.y},
          {
            name: this._algorithm,
            namedCurve: "P-256"
          },
          true,
          []
        )
      }
    } else {
      this.key = await webcrypto.subtle.generateKey({
        name: this._algorithm,
        namedCurve: "P-256"
      },
        true,
        this._usage
      )
    }
    return this
  }
  async sign(message: Uint8Array) {
    if (this.key === undefined) {
      await this.generate();
    }
    if (this.key !== undefined) {
      let signatureBuf = await webcrypto.subtle.sign(
        {
          name: "ECDSA",
          hash: { name: "SHA-256" },
        },
        this.key.privateKey as CryptoKey,
        message
      )
      return new Uint8Array(signatureBuf)
    }
    throw new Error('key is undefined');
  }
  async publicSerial(showName: boolean=true) {
    if (this.key === undefined) {
      await this.generate();
    }
    if (this.key !== undefined) {
      let publicKey = (await webcrypto.subtle.exportKey('raw', this.key.publicKey as CryptoKey)).slice(1);
      let out = (this.name && showName ? `${this.name}.` : '') + b64encode(new Uint8Array(publicKey));
      if (this.repr === undefined) {
        this.repr = out;
      }
      return out;
    }
    throw new Error('key is undefined')
  }
  async exportKey() {
    if (this.exportedKey == undefined) {
      if (this.key === undefined) {
        await this.generate();
      }
      if (this.key !== undefined) {
        this.exportedKey = 
          "-----BEGIN PRIVATE KEY-----\n" +
          b64encode(new Uint8Array(await webcrypto.subtle.exportKey("pkcs8", this.key.privateKey as CryptoKey))) +
          "\n-----END PRIVATE KEY-----" + (this.name? `\n${this.name}` : '')
      } else {
        throw new Error('key is undefined');
      }
    }
    return this.exportedKey;
  }
}
export class PublicKey {
  _algorithm: "ECDSA" | "ECDH";
  _usage: ["verify"] | [];
  publicSerial: string;
  key?: CryptoKey;
  name?: string;

  constructor(use: "verify" | "derive", publicSerial: string) {
    this.publicSerial = publicSerial
    this.name = publicSerial.split('.').slice(0, -1).join('.')
    if (use === 'verify') {
      this._algorithm = 'ECDSA'
      this._usage = ['verify']
      return
    }
    if (use === 'derive') {
      this._algorithm = 'ECDH'
      this._usage = []
      return
    }
    throw new Error("expected input 'use' should be 'verify' or 'derive'");

  }
  async generate() {
    let decodedPublicSerial = b64decode(this.publicSerial.split('\n').slice(-1)[0])
    this.key = await webcrypto.subtle.importKey(
      'raw',
      new Uint8Array([4, ...decodedPublicSerial]).buffer,
      {
        name: this._algorithm,
        namedCurve: 'P-256'
      },
      true,
      this._usage
    )
    return this
  }
  async verify(signature: Uint8Array, message: Uint8Array) {
    if (this.key === undefined) {
      await this.generate()
    }
    if (this.key !== undefined) {
      if (await webcrypto.subtle.verify(
        {
          name: "ECDSA",
          hash: { name: "SHA-256" },
        },
        this.key,
        signature,
        message
      )) { return; } else { throw new Error('Invalid Signature'); }
    }
    throw new Error('key is undefined');
  }
}
export class SharedKey {
  publicSerial: string;
  privateKey: PrivateKey;
  key?: CryptoKey;

  constructor(privateKey: PrivateKey, publicSerial: string) {
    this.publicSerial = publicSerial
    this.privateKey = privateKey
  }
  async generate() {
    if (this.privateKey.key === undefined) {
      await this.privateKey.generate();
    }
    if (this.privateKey.key !== undefined) {
      this.key = await webcrypto.subtle.deriveKey(
        {
          name: "ECDH",
          public: (await new PublicKey('derive', this.publicSerial).generate()).key
        },
        this.privateKey.key.privateKey as CryptoKey,
        {
          name: "AES-CTR",
          length: 256
        },
        false,
        ['encrypt', 'decrypt']
      );
      return this;
    }
  }
  async encrypt(message: Uint8Array, nonce: Uint8Array) {
    if (this.key === undefined) {
      await this.generate()
    }
    if (this.key !== undefined) {
      let out = await webcrypto.subtle.encrypt(
        {
          name: "AES-CTR",
          counter: nonce,
          length: 128
        },
        this.key,
        message
      )
      return new Uint8Array(out);
    }
  }
  async decrypt(ciphertext: Uint8Array, nonce: Uint8Array) {
    if (this.key === undefined) {
      await this.generate()
    }
    if (this.key !== undefined) {
      return await webcrypto.subtle.encrypt(
        {
          name: "AES-CTR",
          counter: nonce,
          length: 128
        },
        this.key,
        ciphertext
      )
    }
  }
}
export class Token {
  issuer: string;
  brokers: string[];
  receiver: string;
  asset: string;
  tokenType: "root" | "extension";
  maxDepth: number | null;
  validFrom: number;
  validUntil: number | null;
  metadata: {};
  signature?: string;

  constructor(
    issuer: string, brokers: string[], receiver: string, asset: string, tokenType: 'root' | 'extension', maxDepth?: number,
    validFrom?: number, validUntil?: number, metadata: {} = {}) {
    this.issuer = issuer;
    this.brokers = brokers;
    this.receiver = receiver;
    this.asset = asset;
    this.tokenType = tokenType;
    this.maxDepth = maxDepth || null;
    this.validFrom = validFrom || Date.now() / 1000;
    this.validUntil = validUntil || null;
    this.metadata = metadata;

  }
  async verify(signature: string) {
    try {
      await new PublicKey('verify', this.issuer).verify(
        b64decode(signature),
        new TextEncoder().encode(this.toString()));
      this.signature = signature;
      return true;
    } catch {
      return false;
    }
  }
  toString() {
    return JSON.stringify(this.toObj())
  }
  toObj() {
    return {
      issuer: this.issuer,
      brokers: this.brokers,
      receiver: this.receiver,
      asset: this.asset,
      token_type: this.tokenType,
      max_depth: this.maxDepth,
      valid_from: this.validFrom,
      valid_until: this.validUntil,
      metadata: this.metadata,
    }
  }
  encode() {
    if (this.signature === undefined) {
      throw 'You need to sign the token before encoding it';
    }
    return this.signature + '.' + this.toString();
  }
  async sign(signingKey: PrivateKey) {
    this.signature = b64encode(await signingKey.sign((new TextEncoder().encode(this.toString()))) as Uint8Array);

    return this.signature
  }
  static async decode(encodedToken: string, verify: boolean = true) {
    let obj = JSON.parse(encodedToken.substr(encodedToken.indexOf('.') + 1));
    let token = new Token(obj.issuer, obj.brokers, obj.receiver, obj.asset, obj.token_type, obj.max_depth, obj.valid_from,
      obj.valid_until, obj.metadata);
    let signature = encodedToken.substring(0, encodedToken.indexOf('.'));
    if (!verify) {
      token.signature = signature;
      return token;
    }
    if (await token.verify(signature)) {
      return token;
    }
    console.error('Invalid Signature');
  }
}